### DISCUSSION TITLE
State management in CeTZ

### DISCUSSION THREAD

**Discussion by noahjutz** [cetz-package/cetz#739] [Q&A] [answered]
Is there a [`state`](https://typst.app/docs/reference/introspection/state/) equivalent in CeTZ? I'd like to create a DSL that allows me to iteratively draw nodes relative to each other, which requires that I keep track of a stack of nodes. It'll look something like this:

```
draw-node[Hello]
down()
draw-node[this is below]
pop()
right()
draw-node[World!]
```

Maybe I can use the Cetz context for this, but I haven't been able to figure out how it works.

**Comment by johannes-wolf** [comment, marked as answer]
Maybe using anchors is enough for drawing nodes next to each other? Note that you can overwrite anchors (e.g. calling something like `anchor("last-elem", (...))` multiple times works).

Yes, you can use cetz context for this. The simplest way to modify the context is by calling `set-ctx(ctx => { ... return ctx })`. If you need more control, you can define your own elements, because all elements can modify the context (note that the context is scoped by groups/scopes).

An element function (like `line()`) returns a single-element array containing a function
that takes a context and returns a dictionary with it's (optional) drawables (paths) and a new context.

For an example, have a look at how `group` or `line` are implemented (in `src/draw/grouping.typ` and `src/draw/shapes.typ`).

**Reply by johannes-wolf** [reply]
Update: There is a PR to support shared-state: #870
